sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java

```diff
@@ -375,4 +422,361 @@ public void checkDone() throws IllegalStateException {
       }
     }
   }
+
+  /**
+   * A splittable {@link DoFn} which executes an {@link UnboundedSource}.
+   *
+   * <p>We model the element as the original source and the restriction as a pair of the sub-source
+   * and its {@link CheckpointMark}. This allows us to split the sub-source over and over as long as
+   * the checkpoint mark is {@code null} or the {@link NoopCheckpointMark} since it does not
+   * maintain any state.
+   */
+  // TODO: Support reporting the watermark, currently the watermark never advances.
```

Comment:
JIRA for tracking? (Or would this be 10897?) Maybe update the PR/commit description to describe the current capabilities and limitations.

Reply:
Will file JIRA if I can't do all the watermark reporting passing and implementation in the SDK harness.